File: output/xCodeEval_remove_8000/C++/Java/3fc93652c831882019e3b86122b70f61.java
import java.io.*;
import java.util.*;

class DSum {
    static class Solver {
        private final BufferedReader in;
        private final PrintWriter out;

        public Solver(BufferedReader in, PrintWriter out) {
            this.in = in;
            this.out = out;
        }

        public void solve() throws IOException {
            StringTokenizer st = new StringTokenizer(in.readLine());
            int N = Integer.parseInt(st.nextToken());
            int K = Integer.parseInt(st.nextToken());

            List<List<Long>> A = new ArrayList<>();
            List<Long> A_sum = new ArrayList<>();

            for (int i = 0; i < N; i++) {
                st = new StringTokenizer(in.readLine());
                int t = Integer.parseInt(st.nextToken());
                List<Long> a = new ArrayList<>();
                for (int j = 0; j < t; j++) {
                    a.add(Long.parseLong(st.nextToken()));
                }
                if (t > K) {
                    a = a.subList(0, K);
                }
                A.add(a);
                long sum = 0;
                for (long num : a) {
                    sum += num;
                }
                A_sum.add(sum);
            }

            long best = 0;
            long[] knapsack = new long[K + 1];
            recurse(0, N, knapsack, A, A_sum, K);

            out.println(best);
        }

        private void add_item(long[] knapsack, int size, long sum) {
            for (int k = knapsack.length - size - 1; k >= 0; k--) {
                knapsack[k + size] = Math.max(knapsack[k + size], knapsack[k] + sum);
            }
        }

        private void recurse(int start, int end, long[] knapsack, List<List<Long>> A, List<Long> A_sum, int K) {
            if (start >= end) {
                return;
            }

            if (end - start == 1) {
                long sum = 0;
                long curBest = knapsack[K];

                for (int prefix = 1; prefix <= A.get(start).size(); prefix++) {
                    sum += A.get(start).get(prefix - 1);
                    curBest = Math.max(curBest, sum + knapsack[K - prefix]);
                }

                best = Math.max(best, curBest);

                return;
            }

            int mid = (start + end) / 2;
            long[] state = knapsack.clone();

            for (int i = start; i < mid; i++) {
                add_item(state, A.get(i).size(), A_sum.get(i));
            }

            recurse(mid, end, state, A, A_sum, K);
            state = knapsack.clone();

            for (int i = mid; i < end; i++) {
                add_item(state, A.get(i).size(), A_sum.get(i));
            }

            recurse(start, mid, state, A, A_sum, K);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);
        Solver solver = new Solver(in, out);
        solver.solve();
        out.close();
    }
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Go/3fc93652c831882019e3b86122b70f61.go
package main

import (
	"fmt"
	"io"
	"os"
)

func addItems(knapsack []int64, size int, sum int64) {
	for k := len(knapsack) - size; k >= 0; k-- {
		knapsack[k+size] = max(knapsack[k+size], knapsack[k]+sum)
	}
}

func recurse(start, end int, knapsack []int64, A [][]int64, A_sum []int64, best *int64) {
	if start >= end {
		return
	}

	if end-start == 1 {
		sum := int64(0)
		*best = max(*best, knapsack[K])

		for prefix := 1; prefix <= len(A[start]); prefix++ {
			sum += A[start][prefix-1]
			*best = max(*best, sum+knapsack[K-prefix])
		}

		return
	}

	mid := (start + end) / 2
	state := make([]int64, K+1)
	copy(state, knapsack)

	for i := start; i < mid; i++ {
		addItems(state, len(A[i]), A_sum[i])
	}

	recurse(mid, end, state, A, A_sum, best)
	state = make([]int64, K+1)
	copy(state, knapsack)

	for i := mid; i < end; i++ {
		addItems(state, len(A[i]), A_sum[i])
	}

	recurse(start, mid, state, A, A_sum, best)
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func DSum(in io.Reader, out io.Writer) {
	var N, K int
	fmt.Fscan(in, &N, &K)

	A := make([][]int64, N)
	A_sum := make([]int64, N)

	for i := 0; i < N; i++ {
		var t int
		fmt.Fscan(in, &t)
		A[i] = make([]int64, t)

		for j := 0; j < t; j++ {
			fmt.Fscan(in, &A[i][j])
		}

		if t > K {
			A[i] = A[i][:K]
		}

		for _, a := range A[i] {
			A_sum[i] += a
		}
	}

	best := int64(0)
	knapsack := make([]int64, K+1)
	recurse(0, N, knapsack, A, A_sum, &best)

	fmt.Fprintln(out, best)
}

func main() {
	DSum(os.Stdin, os.Stdout)
}
